src/plugins/typeahead_location_operator.py

```diff
@@ -0,0 +1,162 @@
+
+        # setup result list
+        typeadhead_result = {}
+
+        # Make dictionary out of list, so it's easier to query later.
+        record_dict = {}
+        record_dict[self.source_location_column] = record[0]
+        record_dict[self.source_key_column] = record[1]
+
+        # Alias entries in the record_dict for easy use
+        key_value = record_dict[self.source_key_column]
+        location_value = record_dict[self.source_location_column]
+
+        # Feed typeahead with location data
+        data_headers = {"content-type": "application/json"}
+        data_url = f"https://api.data.amsterdam.nl/atlas/typeahead/bag/?q={location_value}"
+        data_response = requests.get(data_url, headers=data_headers, verify=False)
+
+        # TODO: find better solution
+        # If no match, try matching only on the first number of
+        # an address (i.e. 'street 1-100 => street 1')
+        if not data_response.json():
+            location_value = location_value.split("-")[0]
+            data_url = f"https://api.data.amsterdam.nl/atlas/typeahead/bag/?q={location_value}"
+            data_response = requests.get(data_url, headers=data_headers, verify=False)
+
+        self.log.error(
+            f"typeahead status {data_response} "
+            f"gets result {data_response.json()} "
+            f"for value {location_value}"
+        )
+
+        # Get the first result, if there are multiple responses
+        # i.e. address without number or addition leads to more than one result
+        try:
+            typeadhead_result[key_value] = data_response.json()[0]["content"][0] or None
+
+        except IndexError:
+            self.log.error(f"No results for {location_value}")
+
+        return typeadhead_result
+
+    def execute(self, context=None):
+        """look up the geometry where no geometry is present"""
+
+        # get location data without geometry
+        rows = self.get_non_geometry_records()
+
+        # get BAG verblijfsobject ID from typeahead
+        for record in rows:
+
+            get_typeadhead_result = self.get_typeahead_result(record)
+            record_key = None
+            bag_url = None
+            for key, value in get_typeadhead_result.items():
+                record_key = key
+                bag_url = value
+
+            # extract the BAG id from the url, which is the last
+            # series of numbers before the last forward-slash
+            try:
+                bag_id = re.search("(/)([0-9]+)(/)$", bag_url["uri"]).group(2)
+                self.log.error(f"BAG id found for {record_key}: {bag_id}")
+
+            except AttributeError:
+                self.log.error(f"No BAG id found for {record_key}, empty result...")
+                continue
+
+            pg_hook = PostgresHook(postgres_conn_id=self.postgres_conn_id)
+
+            with pg_hook.get_cursor() as cursor:
+
+                # update record with found geometry
+                cursor.execute(
```

Comment:
Hier kan het niet zoveel kwaad, omdat Airflow alleen intern draait, maar deze aanpak met string concatenaties is wel gevoelig voor SQL injections. De DB-API heeft hier een standaard aanpak voor met placeholders (voor PostgreSQL is dat %s). Aan de `execute` geeft je dan een lijst (of tuple) met argumenten mee. De DB0-API beschermt je dan standaard voor SQL injections.

Reply:
En het is potentieel sneller. De database kan dan namelijk gebruik maken van een _prepared statement_ (parsing, query optimization, ed zijn dan reeds gedaan). Elke volgende invocatie van het SQL statement hergebruikt dan het _prepared statement_.

Dat gezegd hebben zou je hier dan waarschijnlijk de context manager voor `cursor` om de for loop moeten leggen.

Reply:
Top, ik heb het aangepast